src/L1/rollup/ScrollChain.sol

```diff
@@ -583,10 +623,83 @@ contract ScrollChain is OwnableUpgradeable, PausableUpgradeable, IScrollChain {
         emit FinalizeBatch(_batchIndex, _batchHash, _postStateRoot, _withdrawRoot);
     }
 
+    /// @inheritdoc IScrollChain
+    function finalizeBundleWithTeeProof(
+        bytes calldata _batchHeader,
+        bytes32 _postStateRoot,
+        bytes32 _withdrawRoot,
+        bytes calldata _teeProof
+    ) external override OnlyProver whenFinalizeNotPaused {
+        // compute pending batch hash and verify
+        (uint256 batchPtr, bytes32 _batchHash, uint256 _batchIndex, ) = _loadBatchHeader(_batchHeader);
+
+        // check if this batch is finalized by zk proof.
+        bytes32 cachedStateRoot = finalizedStateRoots[_batchIndex];
+        if (cachedStateRoot == bytes32(0)) {
+            revert ErrorBatchNotFinalizedWithZKProof();
+        }
+
+        uint256 _finalizedBatchIndex = lastTeeFinalizedBatchIndex;
+        if (_batchIndex <= _finalizedBatchIndex) {
+            revert ErrorBatchIsAlreadyVerified();
+        }
+
+        bytes memory _publicInput = abi.encodePacked(
+            layer2ChainId,
+            uint32(_batchIndex - _finalizedBatchIndex), // numBatches
+            finalizedStateRoots[_finalizedBatchIndex], // _prevStateRoot
+            committedBatches[_finalizedBatchIndex], // _prevBatchHash
+            _postStateRoot,
+            _batchHash,
+            _withdrawRoot
+        );
+
+        uint256 batchVersion = BatchHeaderV0Codec.getVersion(batchPtr);
+        IRollupVerifier(teeVerifier).verifyBundleProof(batchVersion, _batchIndex, _teeProof, _publicInput);
+
+        // check withdraw root and state root match with zk proof
+        if (withdrawRoots[_batchIndex] != _withdrawRoot || cachedStateRoot != _postStateRoot) {
+            unresolvedState.batchIndex = _batchIndex;
+            unresolvedState.stateRoot = _postStateRoot;
+            unresolvedState.withdrawRoot = _withdrawRoot;
+            emit StateMismatch(_batchIndex, _postStateRoot, _withdrawRoot);
+            return;
+        }
+
+        lastTeeFinalizedBatchIndex = _batchIndex;
```

Comment:
Again, I would prefer `**VerifiedBatchIndex`, since either of the two proofs is not considered "finalized" anymore. So either `lastTeeVerifiedBatchIndex`, or `lastFinalizedBatchIndex` (since finalized == zkp + tee) would be better here.